From b9054f849be29b2ccad5d0d924f7db1e57a28748 Mon Sep 17 00:00:00 2001
From: Randolph <randolph@andestech.com>
Date: Fri, 6 Sep 2024 16:54:20 +0800
Subject: [PATCH 24/53] swdtb: fixup compatible strings in cpu local interrupt
 controller (#123)

change the compatible in the cpu local interrupt controller from
"riscv,cpu-intc" to "riscv,cpu-intc", "andestech,cpu-intc"

Reviewed-by: Randolph <randolph@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/u-boot.official/pulls/123
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Ben Zong-You Xie <ben717@andestech.com>
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Co-authored-by: Randolph <randolph@andestech.com>
Co-committed-by: Randolph <randolph@andestech.com>
---
 .../common/swdtb_merger/swdtb_merger.c        |  2 +
 .../common/swdtb_merger/swdtb_merger_fixup.c  | 30 ++++++++++++---
 .../common/swdtb_merger/swdtb_merger_fixup.h  |  1 +
 .../common/swdtb_merger/swdtb_tool.c          | 37 ++++++++++++++++---
 .../common/swdtb_merger/swdtb_tool.h          |  6 ++-
 5 files changed, 64 insertions(+), 12 deletions(-)

diff --git a/board/andestech/common/swdtb_merger/swdtb_merger.c b/board/andestech/common/swdtb_merger/swdtb_merger.c
index be2024ba3db7..d4bbdcd24d80 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger.c
@@ -33,6 +33,8 @@ struct swdtb_table dtb_merge_table[] = {
 	 "change ax25 to ae350",	ae350_fixup},
 	{SWDTB_ENABLE,	COUNT_INFINITE,
 	 "isa string for Linux 6.6",	riscv_isa_fixup},
+	{SWDTB_ENABLE,	COUNT_INFINITE,
+	 "using andestech cpu intc",	cpu_intc_fixup},
 	{SWDTB_ENABLE,	1,
 	 "set l2c cache line",		l2c_cache_line_fixup},
 	{SWDTB_ENABLE,	1,
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
index 65830f17f506..42913c9d92b7 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
@@ -29,8 +29,10 @@ int plicsw_fixup(void *fdt, int node, char *name, char *value)
 {
 	char *old_str = "riscv,plic1";
 	char *new_str = "andestech,plicsw";
+	int type = COMPAT_TYPE_REPLACE;
 
-	return compatible_change(fdt, node, name, value, old_str, new_str);
+	return compatible_change(fdt, node, name, value,
+				 old_str, new_str, type);
 }
 
 /**
@@ -44,8 +46,10 @@ int using_generic_dma(void *fdt, int node, char *name, char *value)
 {
 	char *old_str = "andestech,atcdmac300g";
 	char *new_str = "andestech,atcdmac300";
+	int type = COMPAT_TYPE_REPLACE;
 
-	return compatible_change(fdt, node, name, value, old_str, new_str);
+	return compatible_change(fdt, node, name, value,
+				 old_str, new_str, type);
 }
 
 /**
@@ -59,8 +63,10 @@ int using_generic_sdc(void *fdt, int node, char *name, char *value)
 {
 	char *old_str = "andestech,atfsdc010g";
 	char *new_str = "andestech,atfsdc010";
+	int type = COMPAT_TYPE_REPLACE;
 
-	return compatible_change(fdt, node, name, value, old_str, new_str);
+	return compatible_change(fdt, node, name, value,
+				 old_str, new_str, type);
 }
 
 /**
@@ -78,8 +84,10 @@ int ae350_fixup(void *fdt, int node, char *name, char *value)
 	char *old_str = "andestech,a25";
 #endif
 	char *new_str = "andestech,ae350";
+	int type = COMPAT_TYPE_REPLACE;
 
-	return compatible_change(fdt, node, name, value, old_str, new_str);
+	return compatible_change(fdt, node, name, value,
+				 old_str, new_str, type);
 }
 
 /**
@@ -93,8 +101,10 @@ int plmt_fixup(void *fdt, int node, char *name, char *value)
 {
 	char *old_str = "riscv,plmt0";
 	char *new_str = "andestech,plmt0";
+	int type = COMPAT_TYPE_REPLACE;
 
-	return compatible_change(fdt, node, name, value, old_str, new_str);
+	return compatible_change(fdt, node, name, value,
+				 old_str, new_str, type);
 }
 
 /**
@@ -772,3 +782,13 @@ int andespmu_event_fixup(void *fdt, int node, char *name, char *value)
 
 	return !(RETURN_DONE);
 }
+
+int cpu_intc_fixup(void *fdt, int node, char *name, char *value)
+{
+	char *old_str = "riscv,cpu-intc";
+	char *new_str = "andestech,cpu-intc";
+	int type = COMPAT_TYPE_APPEND;
+
+	return compatible_change(fdt, node, name, value, 
+				 old_str, new_str, type);
+}
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
index 9ad434291ee5..e4acf2f3c727 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
@@ -34,6 +34,7 @@ int using_generic_dma(void *fdt, int node, char *name, char *value);
 int using_generic_sdc(void *fdt, int node, char *name, char *value);
 int ae350_fixup(void *fdt, int node, char *name, char *value);
 int plmt_fixup(void *fdt, int node, char *name, char *value);
+int cpu_intc_fixup(void *fdt, int node, char *name, char *value);
 
 int riscv_isa_fixup(void *fdt, int node, char *name, char *value);
 int andespmu_node_fixup(void *fdt, int node, char *name, char *value);
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.c b/board/andestech/common/swdtb_merger/swdtb_tool.c
index 9daa8d6e6a94..f07b44d19016 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.c
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.c
@@ -38,19 +38,26 @@ bool fdt_data_is_one_string(const char *data, int len)
  * @return 0 if ok, or FDT_ERR... if not.
  */
 int compatible_change(void *fdt, int node, char *name, char *value,
-		      char *old_str, char *new_str)
+		      char *old_str, char *new_str, int type)
 {
 	char *prop_name = "compatible";
 	int ret;
+	int (*comp_fixup)(void *fdt, int nodeoffset, const char *name,
+			  const void *val, int len);
+
+	if (type == COMPAT_TYPE_REPLACE)
+		comp_fixup = fdt_setprop;
+	else
+		comp_fixup = fdt_appendprop;
 
 	if ((strcmp(name, prop_name) == 0) &&
 	    (strcmp(value, old_str) == 0)) {
 		dtb_info("RAW: %s: %s\n", (char *)name, (char *)value);
 
-		ret = fdt_setprop(fdt, node, name, new_str,
-				  strlen(new_str) + 1);
+		ret = comp_fixup(fdt, node, name, new_str,
+				 strlen(new_str) + 1);
 		if (ret) {
-			dtb_error("Error on fdt_setprop: %s\n",
+			dtb_error("Error on comp_fixup: %s\n",
 				  fdt_strerror(ret));
 		}
 
@@ -139,14 +146,32 @@ void show_result(const void *fdt, int node, const char *prop_name, int type)
 {
 #ifdef ANDES_SWDTB_MERGER_CHECK
 	const char *data;
+	int count;
+	int idx;
 
 	data = (char *)fdt_getprop(fdt, node, prop_name, NULL);
 	if (!data)
 		dtb_error("Error on fdt_getprop\n");
 
 	if (type == RESULT_STRING) {
-		/* The data is of string type; just display it. */
-		printf("Result: %s: %s%s%s\n", prop_name, KYEL, data, KNRM);
+		/* The data is of string type; just display it. */		
+		count = fdt_stringlist_count(fdt, node, prop_name);
+
+		if (count == 1) {
+			printf("Result: %s: %s%s%s\n", prop_name,
+			       KYEL, data, KNRM);
+			return;
+		}
+
+		printf("Result: %s: ", prop_name);
+		for(idx = 0; idx < count; idx++) {
+			data = fdt_stringlist_get(fdt, node, prop_name,
+						  idx, NULL);
+			printf("%s%s%s", KYEL, data, KNRM);
+			if (idx != count - 1)
+				printf(", ");
+		}
+		printf("\n");
 	} else {
 		printf("Result: %s: %s%d%s\n",
 		       prop_name, KYEL, fdt32_to_cpu(*(u32 *)data), KNRM);
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.h b/board/andestech/common/swdtb_merger/swdtb_tool.h
index 28184965b02a..266a2dffe4ef 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.h
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.h
@@ -65,8 +65,12 @@ void swdtb_global_initial(void);
 
 bool fdt_data_is_one_string(const char *data, int len);
 
+enum compat_type {
+	COMPAT_TYPE_REPLACE,
+	COMPAT_TYPE_APPEND,
+};
 int compatible_change(void *fdt, int node, char *name, char *value,
-		      char *old_str, char *new_str);
+		      char *old_str, char *new_str, int type);
 
 int dma_type_to_node(void *fdt, int node, char *name, const char *target_node);
 int dma_type_fixup(void *fdt, int node, char *name,
-- 
2.34.1

